webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/ZarrAgglomerateService.scala

```diff
@@ -0,0 +1,291 @@
+              positions.getInt(positions.getIndex.set(Array(nodeIdx, 1))),
+              positions.getInt(positions.getIndex.set(Array(nodeIdx, 2)))
+            )
+          },
+          affinities = ArraySeq.unsafeWrapArray(affinities.getStorage.asInstanceOf[Array[Float]])
+        )
+      }.toFox
+    } yield agglomerateGraph
+
+  def segmentIdsForAgglomerateId(agglomerateFileKey: AgglomerateFileKey,
+                                 agglomerateId: Long)(implicit ec: ExecutionContext, tc: TokenContext): Fox[Seq[Long]] =
+    for {
+      agglomerateToSegmentsOffsets <- openZarrArrayCached(agglomerateFileKey, keyAgglomerateToSegmentsOffsets)
+      agglomerateToSegments <- openZarrArrayCached(agglomerateFileKey, keyAgglomerateToSegments)
+      segmentRange <- agglomerateToSegmentsOffsets.readAsMultiArray(offset = agglomerateId, shape = 2)
+      segmentOffset <- tryo(segmentRange.getLong(0)).toFox
+      segmentCount <- tryo(segmentRange.getLong(1) - segmentOffset).toFox
+      segmentIds <- if (segmentCount == 0)
+        Fox.successful(Array.empty[Long])
+      else
+        agglomerateToSegments
+          .readAsMultiArray(offset = segmentOffset, shape = segmentCount.toInt)
+          .flatMap(MultiArrayUtils.toLongArray(_).toFox)
+    } yield segmentIds.toSeq
+
+  def agglomerateIdsForSegmentIds(agglomerateFileKey: AgglomerateFileKey, segmentIds: Seq[Long])(
+      implicit ec: ExecutionContext,
+      tc: TokenContext): Fox[Seq[Long]] =
+    for {
+      segmentToAgglomerate <- openZarrArrayCached(agglomerateFileKey, keySegmentToAgglomerate)
+      agglomerateIds <- Fox.serialCombined(segmentIds) { segmentId =>
+        mapSingleSegment(segmentToAgglomerate, segmentId)
+      }
+    } yield agglomerateIds
+
+  def positionForSegmentId(agglomerateFileKey: AgglomerateFileKey, segmentId: Long)(implicit ec: ExecutionContext,
+                                                                                    tc: TokenContext): Fox[Vec3Int] =
+    for {
+      segmentToAgglomerate <- openZarrArrayCached(agglomerateFileKey, keySegmentToAgglomerate)
+      agglomerateId <- mapSingleSegment(segmentToAgglomerate, segmentId)
+      agglomerateToSegmentsOffsets <- openZarrArrayCached(agglomerateFileKey, keyAgglomerateToSegmentsOffsets)
+      segmentsRange: MultiArray <- agglomerateToSegmentsOffsets.readAsMultiArray(offset = agglomerateId, shape = 2)
+      agglomerateToSegments <- openZarrArrayCached(agglomerateFileKey, keyAgglomerateToSegments)
+      segmentIndex <- binarySearchForSegment(segmentsRange.getLong(0),
+                                             segmentsRange.getLong(1),
+                                             segmentId,
+                                             agglomerateToSegments)
+      agglomerateToPositions <- openZarrArrayCached(agglomerateFileKey, keyAgglomerateToPositions)
+      position <- agglomerateToPositions.readAsMultiArray(offset = Array(segmentIndex, 0), shape = Array(3, 1))
```

Comment:
Why is the shape here `Array(3, 1)`? In the other cases where shape has 2 dims in this file, the first dim is used for element count, and the second for "entries per element". Here it is the other way around :thinking:

Reply:
Good catch, that looks like a bug! I’ll have closer look.